akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -280,68 +485,282 @@ private[akka] class Shard(
+      passivate(sender(), stopMessage)
+    case msg: ShardQuery                 => receiveShardQuery(msg)
+    case PassivateIdleTick               => stash()
+    case msg: RememberEntityStoreCrashed => rememberEntityStoreCrashed(msg)
+    case msg if extractEntityId.isDefinedAt(msg) =>
+      deliverMessage(msg, sender())
+    case msg =>
+      // shouldn't be any other message types, but just in case
+      log.warning(
+        "Stashing unexpected message [{}] while waiting for remember entities update of starts [{}], stops [{}]",
+        msg.getClass,
+        update.started.mkString(", "),
+        update.stopped.mkString(", "))
+      stash()
 
-    if (state.entities(start.entityId)) {
-      getOrCreateEntity(start.entityId)
-      requester ! ShardRegion.StartEntityAck(start.entityId, shardId)
-    } else {
-      processChange(EntityStarted(start.entityId)) { evt =>
-        getOrCreateEntity(start.entityId)
-        sendMsgBuffer(evt)
-        requester ! ShardRegion.StartEntityAck(start.entityId, shardId)
-      }
-    }
   }
 
-  private def receiveStartEntityAck(ack: ShardRegion.StartEntityAck): Unit = {
-    if (ack.shardId != shardId && state.entities.contains(ack.entityId)) {
-      log.debug("Entity [{}] previously owned by shard [{}] started in shard [{}]", ack.entityId, shardId, ack.shardId)
-      processChange(EntityStopped(ack.entityId)) { _ =>
-        state = state.copy(state.entities - ack.entityId)
-        messageBuffers.remove(ack.entityId)
+  def onUpdateDone(starts: Set[EntityId], stops: Set[EntityId]): Unit = {
+    // entities can contain both ids from start/stops and pending ones, so we need
+    // to mark the completed ones as complete to get the set of pending ones
+    starts.foreach { entityId =>
+      val stateBeforeStart = entities.entityState(entityId)
+      // this will start the entity and transition the entity state in sessions to active
+      getOrCreateEntity(entityId)
+      sendMsgBuffer(entityId)
+      stateBeforeStart match {
+        case RememberingStart(ackTo) => ackTo.foreach(_ ! ShardRegion.StartEntityAck(entityId, shardId))
+        case _                       =>
       }
+      touchLastMessageTimestamp(entityId)
+    }
+    stops.foreach { entityId =>
+      entities.entityState(entityId) match {
+        case RememberingStop =>
+          // this updates entity state
+          passivateCompleted(entityId)
+        case state =>
+          throw new IllegalStateException(
+            s"Unexpected state [$state] when storing stop completed for entity id [$entityId]")
+      }
+    }
+
+    val (pendingStarts, pendingStops) = entities.pendingRememberEntities()
+    if (pendingStarts.isEmpty && pendingStops.isEmpty) {
+      if (verboseDebug) log.debug("Update complete, no pending updates, going to idle")
+      unstashAll()
+      context.become(idle)
+    } else {
+      // Note: no unstashing as long as we are batching, is that a problem?
```

Comment:
?

Reply:
We discussed this a bit, me and @chbatey, if it is a problem wrt fairness that if you continously keep getting new write batches while writing the stashed messages are delayed until that spike of starting/stopping has completed. We ended up thinking it isn't a problem but perhaps we were wrong.

Reply:
I should add that way less messages are ending up in the stash now than before while writing.